### PR TITLE
cmd/gitserver: add p4-fusion to gitserver Docker image

### PR DESCRIPTION
**Pull request by sashaostrikov**
Included p4-fusion built from salesforce sources to gitserver Docker image

Closes https://github.com/sourcegraph/sourcegraph/issues/26926

### CODE REVIEW

File: cmd/gitserver/Dockerfile [hunk 1]
```diff
@@ -19,6 +19,40 @@ RUN wget -O coursier.zip https://github.com/sourcegraph/lsif-java/releases/downl
...
+	libarchive && \
+    apk update
+
+# Fetching p4 sources archive
+RUN wget https://github.com/salesforce/p4-fusion/archive/refs/tags/v1.2.zip && \
+    mv v1.2.zip /usr/local/bin && \
+    unzip /usr/local/bin/v1.2.zip -d /usr/local/bin
+
+# We need a specific version of OpenSSL
+RUN curl https://www.openssl.org/source/openssl-1.0.2t.tar.gz | tar xz && cd openssl-1.0.2t && ./config && make && make install
```
Comment (Strum355) [inline cmd/gitserver/Dockerfile:42]: Do we know if openssl is statically linked? If not, that might be an issue if we need a specific openssl version but the final image has a different one

Comment (sashaostrikov) [reply cmd/gitserver/Dockerfile:42]: I can tell that it is statically linked. All versions on [downloads page](https://www.openssl.org/source/old/1.0.2/) have a dedicated URL for them, so I guess they should not change at any time in the future

Comment (Strum355) [reply cmd/gitserver/Dockerfile:42]:
```
ldd /usr/local/bin/p4-fusion-1.2/build/p4-fusion/p4-fusion
	/lib/ld-musl-x86_64.so.1 (0x7f407c065000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x7f407b99a000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x7f407b986000)
	libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x7f407c065000)
```
seems fine :+1:

**Review by Strum355 — APPROVED** [review]

**Comment by ryanslade** [comment]
Looks like some unrelated changes are causing the build to fail, probably worth merging `main` back in

**Comment by ryanslade** [comment]
@sashaostrikov Looks like there's a linter error to sort out now :)

**Comment by ryanslade** [comment]
I tried building the image as explained in the issue. I think started a sh and tried to run `p4-fusion` but I get a bunch of errors like this: 

```
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/bin/p4-fusion: _ZTISt9bad_alloc: symbol not found
Error relocating /usr/local/bin/p4-fusion: _ZTVN10__cxxabiv121__vmi_class_type_infoE: symbol not found
Error relocating /usr/local/bin/p4-fusion: _ZTVN10__cxxabiv121__vmi_class_type_infoE: symbol not found
Error relocating /usr/local/bin/p4-fusion: _ZTVN10__cxxabiv121__vmi_class_type_infoE: symbol not found
Error relocating /usr/local/bin/p4-fusion: _ZSt4cerr: symbol not found
Error relocating /usr/local/bin/p4-fusion: _ZSt4cout: symbol not found
```

It looks like to me like their is an issue linking something.

**Comment by sashaostrikov** [comment]
> I tried building the image as explained in the issue. I then started a sh and tried to run `p4-fusion` but I get a bunch of errors like this:
> 
> ```
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: __cxa_pure_virtual: symbol not found
> Error relocating /usr/local/bin/p4-fusion: _ZTISt9bad_alloc: symbol not found
> Error relocating /usr/local/bin/p4-fusion: _ZTVN10__cxxabiv121__vmi_class_type_infoE: symbol not found
> Error relocating /usr/local/bin/p4-fusion: _ZTVN10__cxxabiv121__vmi_class_type_infoE: symbol not found
> Error relocating /usr/local/bin/p4-fusion: _ZTVN10__cxxabiv121__vmi_class_type_infoE: symbol not found
> Error relocating /usr/local/bin/p4-fusion: _ZSt4cerr: symbol not found
> Error relocating /usr/local/bin/p4-fusion: _ZSt4cout: symbol not found
> ```
> 
> It looks like to me like their is an issue linking something.

can it be that some of the sources should be located nearby to make this work properly?
I'll try copying the sources to the image as well and see how it will work

**Comment by sashaostrikov** [comment]
@ryanslade I found out that this issue is because of alpine linux Docker image, there is a [workaround](https://stackoverflow.com/questions/53875259/error-relocating-symbol-not-found-building-docker-fpm-alpine-image) to change `alpine main` repository to `alpine testing` within the image. I am trying to do this trick and see how it will work

**Comment by ryanslade** [comment]
> @ryanslade I found out that this issue is because of alpine linux Docker image _which uses a non standard musl c runtime library_, there is a [workaround](https://stackoverflow.com/questions/53875259/error-relocating-symbol-not-found-building-docker-fpm-alpine-image) to change `alpine main` repository to `alpine testing` within the image. I am trying to do this trick and see how it will work

Strange, it worked before with this [Dockerfile](https://github.com/sourcegraph/p4-cpp-client/blob/ed56a36b0121a63bb48f2373c9d180c3eaf59ee8/Dockerfile). But that was while back and the code may have changed quite a bit since then.

**Comment by sashaostrikov** [comment]
@ryanslade now p4-fusion can be launched successfully

File: cmd/gitserver/Dockerfile [hunk 2]
```diff
@@ -37,18 +75,32 @@ RUN apk add --no-cache \
...
+
 # This is a trick to include libraries required by p4,
 # please refer to https://blog.tilander.org/docker-perforce/
 ADD https://github.com/jtilander/p4d/raw/4600d741720f85d77852dcca7c182e96ad613358/lib/lib-x64.tgz /
 RUN tar zxf /lib-x64.tgz --directory /
 
+# We need a specific version of OpenSSL to run p4-fusion
+RUN wget https://www.openssl.org/source/openssl-1.0.2t.tar.gz && tar -xzvf openssl-1.0.2t.tar.gz
+
+WORKDIR /openssl-1.0.2t
```
Comment (ryanslade) [inline cmd/gitserver/Dockerfile:100]: I have no idea if it matters, but I think the container will be left having this as the default directory when we hit the entrypoint lower down.

Maybe it's safer to set `WORDIR /` right before the `ENTRYPOINT`?

**Review by ryanslade — APPROVED** [review]

**Comment by sashaostrikov** [comment]
@ryanslade 
fixed workdir
PTAL

**Review by ryanslade — APPROVED** [review]
LGTM!

**Comment by stefanhengl** [comment]
@sashaostrikov @ryanslade Just saw that the gitserver Dockerfile is now in an inconsistent state. This PR uses an older alpine version and an older git to build p4-fusion. Are the older versions somehow required to build p4-fusion or should we update before the release cut?

**Comment by ryanslade** [comment]
Nope, we should be fine to use the latest version.

**Comment by ryanslade** [comment]
And yes, we should update. I realise my "Nope" above could be confusing 😄

**Comment by ryanslade** [comment]
I'll create a PR

**Comment by ryanslade** [comment]
Strange, I can use the latest alpine without error but if I try and use `git 2.34.1` in the `p4-fusion` builder image it build without error, but when I try and run it I get this error: 

```
/ $ p4-fusion
Error relocating /usr/local/bin/p4-fusion: qsort_r: symbol not found
```

**Comment by sashaostrikov** [comment]
@ryanslade if you revert git version, it runs successfully?

**Comment by ryanslade** [comment]
> @ryanslade if you revert git version, it runs successfully?

Yes, if I revert in the builder image but keep the latest version in the final build it all seems to work. My concern is we may find that it breaks at runtime in a different code path. 

My guess is that the vendored version of libgit2 might not work with the latest version of git. They've vendored libgit2 version 1.1.1 and the latest release is 1.3.0

**Comment by sashaostrikov** [comment]
> > @ryanslade if you revert git version, it runs successfully?
> 
> Yes, if I revert in the builder image but keep the latest version in the final build it all seems to work. My concern is we may find that it breaks at runtime in a different code path.
> 
> My guess is that the vendored version of libgit2 might not work with the latest version of git. They've vendored libgit2 version 1.1.1 and the latest release is 1.3.0

I think that's the moment when we can ask developers of Salesforce for potential libgit2 compatibility issues

**Comment by ryanslade** [comment]
I think we may want to revert our changes to the `Dockerfile` then since it's unlikely we'll find a quick solution today and a release is being cut later.

**Comment by sashaostrikov** [comment]
> I think we may want to revert our changes to the `Dockerfile` then since it's unlikely we'll find a quick solution today and a release is being cut later.

Do you mean changes in scope of [this PR](https://github.com/sourcegraph/sourcegraph/pull/28905)?

**Comment by ryanslade** [comment]
Yep. It's not super urgent so it's safe to roll back for now. The client is happy to test with insider builds for now which are our latest builds from CI so they don't need to wait until the next release to test this.

**Comment by ryanslade** [comment]
Something weird going on here.. With a single `Dockerfile` just building `p4-fusion` with latest alpine and git works fine:

```Dockerfile
FROM sourcegraph/alpine-3.12:120059_2021-12-09_b34c7b2@sha256:9a1fde12f56fea02027cf4caeebdddfedb7b73bf8db6c16f7907a6e04a29134c as p4-fusion

# hadolint ignore=DL3018
RUN apk add --no-cache \
	# We require git 2.34.1 because we use git-repack with flag --write-midx.
	'git=~2.34.1' --repository=http://dl-cdn.alpinelinux.org/alpine/v3.15/main  \
	make \
	cmake \
	g++ \
	gcc \
	perl \
	libarchive

RUN apk update 

WORKDIR /app

COPY . . 

# We need a specific version of OpenSSL
RUN curl https://www.openssl.org/source/openssl-1.0.2t.tar.gz | tar xz && cd openssl-1.0.2t && ./config && make && make install 

# Build p4-fusion
RUN ./generate_cache.sh Release
RUN ./build.sh

USER sourcegraph
```

**Pull request merged**